sycl/doc/extensions/supported/sycl_ext_intel_esimd/examples/dpas.md

```diff
@@ -0,0 +1,268 @@
+  mask = mask << elem_bit_offset;
+  target_elem = (target_elem & ~mask) | i_value;
+  vec[packed_linear_index] = target_elem;
+}
+
+BPackedType read_from_vertically_packed_matrix_b(BPackedType *vvec, int row,
+                                                 int col) {
+  constexpr int NumRows = K;
+  constexpr int NumCols = N;
+
+  int *vec = reinterpret_cast<int *>(vvec);
+
+  // Assume that we deal only with 2 or 4-bit integers.
+  static_assert((BElemBitSize == 2 || BElemBitSize == 4),
+                "Unexpected element type");
+  // Assume the packed type is unsigned.
+  static_assert(std::is_unsigned_v<BPackedType>, "Expect unsigned packed type");
+
+  // Find and read the target 'int' element.
+  // The unpacked matrix has dimensions: NumRows*NumCols.
+  constexpr int ElemsInInt = 32 / BElemBitSize;
+
+  int packed_row = row / ElemsInInt;
+  int target_elem = vec[packed_row * NumCols + col];
+
+  // 2. Extract the queried sub-elem from 32-bit int, bit-cast to BPackedType
+  // and return.
+  int elem_bit_offset = (row % ElemsInInt) * BElemBitSize;
+  unsigned int mask = (static_cast<uint64_t>(1) << BElemBitSize) - 1;
+  int value = (target_elem >> elem_bit_offset) & mask;
+  return static_cast<BPackedType>(value);
+}
+
+int main() {
+  unsigned n_errs = 0;
+  try {
+    queue q(gpu_selector_v, create_exception_handler());
+    auto dev = q.get_device();
+    std::cout << "Running on " << dev.get_info<info::device::name>()
+              << std::endl;
+
+    constexpr unsigned Size = 128;
+    constexpr unsigned VL = 16;
+
+    constexpr int APackedSize =
+        M * K * AElemBitSize / (sizeof(APackedType) * 8);
+    constexpr int BPackedSize =
+        K * N * BElemBitSize / (sizeof(BPackedType) * 8);
+
+    auto a_packed = aligned_alloc_shared<APackedType>(128, APackedSize, q);
+    auto b_packed = aligned_alloc_shared<BPackedType>(128, BPackedSize, q);
+    auto res = aligned_alloc_shared<ResType>(128, M * N, q);
+
+    std::unique_ptr<APackedType, usm_deleter> guard_a(a_packed, usm_deleter{q});
+    std::unique_ptr<BPackedType, usm_deleter> guard_b(b_packed, usm_deleter{q});
+    std::unique_ptr<ResType, usm_deleter> guard_res(res, usm_deleter{q});
+
+    // Initialize a_packed;
+    unsigned value = 0;
+    for (int i = 0; i < M; i++) {
+      for (int j = 0; j < K; j++) {
+        value += 1;
+        write_to_horizontally_packed_matrix_a(a_packed, i, j,
+                                              static_cast<APackedType>(value));
+      }
+    }
+
+    // Initialize b_packed;
+    for (int i = 0; i < K; i++) {
+      for (int j = 0; j < N; j++) {
+        int value = (i + j % 4) == 0 ? 1 : (2 + i + j) % 3;
+        write_to_vertically_packed_matrix_b(b_packed, i, j,
+                                            static_cast<BPackedType>(value));
+        assert(value == (int)(static_cast<BPackedType>(value)) && "ERROR");
+      }
+    }
+
+    q.single_task([=]() SYCL_ESIMD_KERNEL {
```

Comment:
In the only in-tree example of an ESMID kernel we are using `[[intel::sycl_explicit_simd]]`, so maybe we should use that for consistency?

Reply:
Yes, makes sense.